### PR TITLE
feat(gtm): add optionnal consent mode

### PR DESCRIPTION
**Pull request by huang-julien**
linked to https://github.com/nuxt/scripts/issues/243
same as https://github.com/GoogleChromeLabs/third-party-capital/pull/71 but for GTM

Seems like this require to push `arguments` into the datalayer array. 🤷‍♂️ 

Result with the GTM debugger:

image

### CODE REVIEW

**Review by housseindjirdeh — APPROVED** [review]
LGTM. Will defer to @flashdesignory for a sanity check before we merge

**Review by flashdesignory — APPROVED** [review]
🚀

File: data/google-tag-manager.json
Comment (felixarntz) [inline data/google-tag-manager.json:18]: Are we certain that including this in every Analytics `gtag` snippet is safe (i.e. leads to the same behavior as if it wasn't included)? I wonder because it's not mentioned in the default Analytics snippet.

Without being certain about this, I'd feel more comfortable if we could make inclusion of this call conditional, e.g. only include the `consent` call if the `consentType` and `consentValue` argument are specified/not empty, and then leave the defaults empty and up to the consumer.

Comment (flashdesignory) [reply data/google-tag-manager.json:18]: hmm yeah, that's a valid point. 
Could we use one optionalParam ("consent") and default it to null?
In that way it wouldn't get included unless you pass in an object?

Comment (flashdesignory) [reply data/google-tag-manager.json:18]: @housseindjirdeh @huang-julien

Comment (huang-julien) [reply data/google-tag-manager.json:18]: default everything to null ? What would happen if we call gtm with nothing ? Will it ignore it ?

Comment (felixarntz) [reply data/google-tag-manager.json:18]: It's not required to provide any `consent` to GTM, so not making the `consent` call would work fine. For instance, the default GTM snippets don't include that portion either (see https://support.google.com/tagmanager/answer/14847097 for example).

Comment (huang-julien) [reply data/google-tag-manager.json:18]: I guess we probably want to have `(function () {window[{{l}}].push(arguments)})('consent', {{consentType}},{{consentValues}})` conditionnally. 

The best thing to do is probbly to provide a `preConfigCode` optionnal param so user will be even more free to add what they want. THe only issue is that `formatCode` stringify strings before insertion so we won't be able to add code with replacement tokens.

Comment (felixarntz) [reply data/google-tag-manager.json:18]: I think it'd be safer to avoid direct code injection, so allowlisting certain kinds of parameters would IMO be a better option. I think the current approach of conditional `consentType` and `consentValues` works well, or alternatively we could use just a single `consent` key which includes both (e.g. in an array).

There aren't many of those special use-cases where configuration is needed _prior_ to the config line. The only other one I'm aware of is the `gtag('set', ...)` command (see #68, related: #71), which we could also support with an optional parameter (let's leave that for a separate issue/PR though).

Comment (felixarntz) [reply data/google-tag-manager.json:18]: Related: With the approach from #73, we would be able to conditionally pass the consent call here too, leaving the defaults as it is today.

**Comment by huang-julien** [comment]
Setting this to draft mode until #73 lands on main branch

**Comment by felixarntz** [comment]
@huang-julien Just merged #73, so you should now be able to use the `{{#consentValues}}...{{/consentValues}}` syntax from there.

**Comment by huang-julien** [comment]
Woohoo 🎉🎉 thank you ! I'll update the PR tomorrow !

File: data/google-tag-manager.json
Comment (felixarntz) [inline data/google-tag-manager.json:21]: I think `consentValues` needs to be present here, like with Analytics? https://github.com/GoogleChromeLabs/third-party-capital/blob/main/data/google-analytics.json#L23
```suggestion
          "consentType": "default",
          "consentValues": null
```

File: src/types/type-declarations.ts
Comment (felixarntz) [inline src/types/type-declarations.ts:109]:
```suggestion
   * @default null
```

**Review by felixarntz — COMMENTED** [review]
@huang-julien Looks good except two things, I think specifically the inclusion of `consentValues` in the schema `optionalParams` is important.

Once this is merged, the PHPUnit tests will probably fail, I can follow up with that. I just realized the reason that this isn't flagged here is because we have an allowlist of file types in there, and the `data/*.json` files aren't among them - so we should fix so that in the future we'll find out "automatically" prior to merging.

File: data/google-tag-manager.json
Comment (huang-julien) [reply data/google-tag-manager.json:21]: oh, consentValues can be undefined. null and undefined both count as falsy boolean expression. We can also remove it from analytics so the JSON can be slightly lighter.

File: src/types/type-declarations.ts
Comment (huang-julien) [reply src/types/type-declarations.ts:109]:
```suggestion
   * @default undefined
```

Comment (huang-julien) [reply src/types/type-declarations.ts:109]: it's already undefineable. We can remove this line

File: data/google-tag-manager.json
Comment (felixarntz) [reply data/google-tag-manager.json:21]: > oh, consentValues can be undefined. null and undefined both count as falsy boolean expression. We can also remove it from analytics so the JSON can be slightly lighter.

Generally makes sense, however wouldn't it be more intuitive to still have those be explicitly listed in optionalParams? I think it helps DX to be able to quickly scan the params and optionalParams and know what's available. Otherwise you'll have to look at the less obvious and harder to parse "code" value.

I don't think it technically makes a difference whether they're explicitly defined or not, but personally I think there's a benefit of keeping them in there.

Comment (huang-julien) [reply data/google-tag-manager.json:21]: Does PHP unit test needs some changes ?

Comment (huang-julien) [reply data/google-tag-manager.json:21]: I get it now, llet's keep it :+1:

**Review by felixarntz — APPROVED** [review]

**Pull request merged**
